lib/closure_tree/hierarchy_maintenance.rb

```diff
@@ -94,19 +94,43 @@ def delete_hierarchy_references
         # It shouldn't affect performance of postgresql.
         # See http://dev.mysql.com/doc/refman/5.0/en/subquery-errors.html
         # Also: PostgreSQL doesn't support INNER JOIN on DELETE, so we can't use that.
-        _ct.connection.execute <<-SQL.squish
+        if mysql?
+          _ct.connection.execute <<-SQL.squish
+            DELETE #{_ct.quoted_hierarchy_table_name} FROM #{_ct.quoted_hierarchy_table_name}
+              JOIN (
+                SELECT descendant_id FROM #{_ct.quoted_hierarchy_table_name}
+                WHERE ancestor_id = #{_ct.quote(id)}
+                    OR descendant_id = #{_ct.quote(id)}
+              ) #{_ct.t_alias_keyword} x
+              ON x.descendant_id = #{_ct.quoted_hierarchy_table_name}.descendant_id
+          SQL
+        else
+          _ct.connection.execute <<-SQL.squish
           DELETE FROM #{_ct.quoted_hierarchy_table_name}
           WHERE descendant_id IN (
             SELECT DISTINCT descendant_id
             FROM (SELECT descendant_id
               FROM #{_ct.quoted_hierarchy_table_name}
               WHERE ancestor_id = #{_ct.quote(id)}
                  OR descendant_id = #{_ct.quote(id)}
-            ) #{ _ct.t_alias_keyword } x )
-        SQL
+            ) #{_ct.t_alias_keyword} x )
+          SQL
+        end
       end
     end
 
+    def mysql?
+      env_db == :mysql2
```

Comment:
Hmm. This is actually not a great test, there are other adapters for mysql. We're planning on switching to `trilogy` within the next few months :-\

Reply:
I'm not actually sure if there _is_ a good way to determine that the underlying database is mysql though. Perhaps instead, we should be using a supplied configuration option, like `apply_mysql_optimizations = true`?

Reply:
Sorry - ignore my comment. This code is way beyond db-agnosticism.

Reply:
> Were you intending to make a PR into the `master` branch of _our fork_ of `clojure_tree`, or were you trying to upstream a solution? If you're just trying to see if it _solves the problem_, then I think you should just build a prerelease version locally that doesn't even worry about it maybe being postgres and push that to gemfury to use.
> 
> I'd really recommend against trying to maintain a long-term fork though, if that was the plan. Instead, we can shim a mysql-specific solution on there that _can't_ be upstreamed, and start a conversation with the gem maintainers about how to fix the real problem properly. If you take that approach, you don't need a `mysql?` check, you can just merge mysql-specific code, alongside a comment explaining what you're doing. But we'll need a prerelease version number and to switch to loading from gemfury.

No, I'm not intending to make a PR into the `master` of our fork. I only opened the PR to get feedback and then after that reference the branch like we're doing with other forks in orders like [this one](https://github.com/usertesting/orders/blob/master/Gemfile#L52). Let me know if that is not the recommended approach anymore in orders. I'm also happy to use gemfury instead :)

This is a short-term fork since we're only seeing this problem because we're on MySQL 5.7. When we upgrade to 8.0 we'll then go back to using the upstream's master branch.

Reply:
Oh, thanks for pointing that out, I did not know about the `trilogy ` plan. We do not need the conditional at all, to be honest. I'm only keeping it to make sure the specs in the gem pass and I'm not introducing any bugs :) Only `sqlite3` does not like the delete join syntax. Isolating sqlite in the `if` condition with the new code in the `else` block makes everything pass and I'll use that instead.